### PR TITLE
Progress for anonymous users 2

### PR DESCRIPTION
**Pull request by mehalshah**

### CODE REVIEW

File: dashboard/app/assets/javascripts/client_state.js [hunk 1]
```diff
@@ -41,16 +49,33 @@ dashboard.clientState.levelProgress = function(level) {
...
+ * @param {number} lines - Number of lines of code user wrote in this solution
+ * @param {number} testResult - Indicates pass, fail, perfect
+ * @param {number} scriptLevelId - Which level this is for
+ */
+dashboard.clientState.trackProgress = function(result, lines, testResult, scriptLevelId) {
+  if (result) {
+    addLines(lines);
+  }
+
+  if (testResult > dashboard.clientState.levelProgress(scriptLevelId)) {
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/client_state.js:63]: By the way, if it's annoying to write dashboard.clientState.levelProgress, you could just define "function levelProgress" inside the closure and then assign " dashboard.clientState.levelProgress = levelProgress". Not important if you don't mind though, either way is fine.

File: dashboard/app/assets/javascripts/client_state.js [hunk 2]
```diff
@@ -25,6 +25,14 @@ dashboard.clientState = {};
  */
 dashboard.clientState.EXPIRY_DAYS = 365;
 
+/**
+ * Maximum number of lines of code that can be stored in the cookie
+ * @type {number}
+ * @private
+ */
+dashboard.clientState.MAX_LINES_TO_SAVE = 1000;
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/client_state.js]: Is this an internal implementation detail?  You could just make it a variable inside the closure.

File: dashboard/app/controllers/script_levels_controller.rb [hunk 3]
```diff
@@ -152,6 +152,10 @@ def present_level
       has_i18n: @game.has_i18n?
     )
 
+    level_view_options(
+      script_level_id: @script_level.level_id
```
Comment (philbogle) [inline dashboard/app/controllers/script_levels_controller.rb:156]: Where is the new script_level_id field used?

File: dashboard/spec/javascripts/client_state_test.js [hunk 4]
```diff
@@ -1,63 +1,79 @@
...
-    $.cookie('progress', '',  {expires: 365})
-    var state = dashboard.clientState;
-    state.levelProgress(100).should.equal(0);
+    //User has passed a level with optimal solution
+    state.trackProgress(true, 5, 100, 1);
+    state.levelProgress(1).should.equal(100);
+    state.levelProgress(2).should.equal(0);
+    state.lines().should.equal(5);
 
-    $.cookie('progress', '[(*', {expires: 365})
```
Comment (philbogle) [inline dashboard/spec/javascripts/client_state_test.js:28]: Could we preserve the malformed cookie test?  Good to know we handle this gracefully.

**Comment by philbogle** [comment]
I believe our existing UI tests cover the updated functionality; is that right?  (OK not to test the too many lines functionality, that's a corner case.)

LGTM after comments are addressed.

File: dashboard/app/assets/javascripts/client_state.js (hunk 2 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/client_state.js]: Okay

File: dashboard/app/controllers/script_levels_controller.rb (hunk 3 above)
Comment (mehalshah) [inline dashboard/app/controllers/script_levels_controller.rb:156]: SendReport.js. The name is converted from script_level_id to scriptLevelId

File: dashboard/spec/javascripts/client_state_test.js (hunk 4 above)
Comment (mehalshah) [inline dashboard/spec/javascripts/client_state_test.js:28]: Will add

File: dashboard/app/controllers/script_levels_controller.rb (hunk 3 above)
Comment (philbogle) [inline dashboard/app/controllers/script_levels_controller.rb:156]: OK, got it, makes sense.

**Pull request merged**